### PR TITLE
mgr/cephadm: block draining last _admin host

### PR DESCRIPTION
**Pull request by adk3798**
Fixes: https://tracker.ceph.com/issues/54413

Signed-off-by: Adam King <adking@redhat.com>

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [x] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/pybind/mgr/orchestrator/_interface.py [hunk 1]
```diff
@@ -355,7 +355,7 @@ def remove_host(self, host: str, force: bool, offline: bool) -> OrchResult[str]:
         """
         raise NotImplementedError()
 
-    def drain_host(self, hostname: str) -> OrchResult[str]:
+    def drain_host(self, hostname: str, force: bool) -> OrchResult[str]:
```
Comment (nizamial09) [inline src/pybind/mgr/orchestrator/_interface.py:358]: @adk3798 I think this broke the API call from dashboard. That's why cephadm e2e failed. So can we default the `force` to `False`?
```suggestion
    def drain_host(self, hostname: str, force: bool = False) -> OrchResult[str]:
```

Comment (adk3798) [reply src/pybind/mgr/orchestrator/_interface.py]: done

**Review by nizamial09 — APPROVED** [review]

**Comment by adk3798** [comment]
http://pulpito.front.sepia.ceph.com/adking-2022-02-28_06:27:32-orch:cephadm-wip-adk-testing-2022-02-27-1813-distro-basic-smithi/

green

File: src/pybind/mgr/cephadm/module.py [hunk 2]
```diff
@@ -2701,12 +2701,26 @@ def remove_osds_status(self) -> List[OSD]:
...
-    def drain_host(self, hostname):
-        # type: (str) -> str
+    def drain_host(self, hostname, force=False):
+        # type: (str, bool) -> str
         """
         Drain all daemons from a host.
         :param host: host name
         """
+
+        # if we drain the last admin host we could end up remoing the only instance
```
Comment (mgfritch) [inline src/pybind/mgr/cephadm/module.py:2711]:
```suggestion
        # if we drain the last admin host we could end up removing the only instance
```
(nit)

Comment (phlogistonjohn) [inline src/pybind/mgr/cephadm/module.py:2711]: Typo: "remoing" should be "removing"

**Review by mgfritch — APPROVED** [review]
covers the `host drain` use-case, but we still have an issue if the user were to remove the label (`host label rm`) from the last _admin host...

**Comment by adk3798** [comment]
> covers the `host drain` use-case, but we still have an issue if the user were to remove the label (`host label rm`) from the last _admin host...
> 
> see: https://tracker.ceph.com/issues/54425

ACK, I'll get something for this case in a follow up PR

File: src/pybind/mgr/cephadm/module.py (hunk 2 above)
Comment (adk3798) [reply src/pybind/mgr/cephadm/module.py]: oops

**Comment by adk3798** [comment]
http://pulpito.front.sepia.ceph.com/adking-2022-03-01_13:09:22-orch:cephadm-wip-adk-testing-2022-02-28-2049-distro-basic-smithi/

2 unrelated failures during ceph-volume commands that will need future investigation
1 https://tracker.ceph.com/issues/54329

**Comment by adk3798** [comment]
jenkins retest this please

**Comment by adk3798** [comment]
jenkins test api

**Comment by ljflores** [comment]
jenkins retest this please

**Comment by ljflores** [comment]
jenkins test make check

**Comment by ljflores** [comment]
jenkins test make check

**Comment by ljflores** [comment]
jenkins test docs

**Comment by ljflores** [comment]
jenkins test docs

**Comment by ljflores** [comment]
jenkins test docs

**Comment by ljflores** [comment]
@adk3798 good news, I got make check to pass! I figured out that there are a handful of jenkins nodes that were causing problems (braggi15, irvingi03, irvingi04, irvingi05, and irvingi07). @tchaikov narrowed it down to the introduction of gcc-11. Not all nodes are affected by this issue (adami05, adami08, and braggi14), so if you re-trigger make check to run on one of those nodes, it will pass. Also, Kefu is implementing a workaround to fix the affected nodes.

**Pull request merged**